Marshaling: beyond 4 Gb

xavierleroy opened the pull request:
Following up on http://caml.inria.fr/mantis/view.php?id=6910, here is a proposal to enable marshaling (output_value, input_value, etc) to process really large data structures.  Currently, the data format used for marshaling limits the size of marshaled data to 4 Gb, owing to the use of 32-bit integers in the data header.  

This pull request introduces an alternate format for the headers of marshaled data, where all sizes are represented as 64 bit integers, thus lifting this limitation.

output_value et al use the alternate, "big" format only if:
- the size of marshaled data would not fit the old, "small" format (which implies we are on a 64-bit platform, obviously);
- the Compat_32 flag is not selected (such "big" marshaled data cannot be read back on a 32-bit platform, obviously).  If Compat_32 is selected, marshaling fails.

In all other cases, we keep using the old, "small" format, for compatibility with 32-bit readers, with older versions of OCaml, and also for compactness (the "small" header is smaller than the "big" header).

input_value et al handle both header formats on 64-bit hosts, and fail cleanly on the big header format on 32-bit hosts.

This PR is a straightforward implementation of this approach.  There is some code refactoring so that the generation and parsing of headers is not duplicated.    In parallel, there are a few independent improvements to the marshaling code:
- replace ugly macros with inline functions (now that we are allowed to use them);
- use memcpy() instead of memmove() when we know the source and destination blocks are disjoint.

I have a simple test that is not included in the PR, however, because it needs about 64 Gb of RAM (yes, you need that much space to exercise the new "big" format) and takes 10+ minutes to run (yes, you need that much time to generate, process and check that much data).  So, this test is not appropriate for inclusion in the test suite.

The most questionable thing about this PR is its practical utility...  As the figures above suggest, one needs a big machine, big data, and big computations to hit the size limitations of the current, "small" header format.  At best, this work is future-proofing.

xavierleroy committed:
Experiment: output_value/input_value beyond the 4 Gb barrier.


git-svn-id: http://caml.inria.fr/svn/ocaml/branches/big-marshal@16323 f963ae5c-01c2-4b8c-9fe0-0dff7051ff02

xavierleroy committed:
Add alternate marshaling format, used when marshaled data is >= 4 Gb.


git-svn-id: http://caml.inria.fr/svn/ocaml/branches/big-marshal@16324 f963ae5c-01c2-4b8c-9fe0-0dff7051ff02

yakobowski commented:
Unless I'm mistaken, a certain startup dedicated to verifying C programs already hit the current limit. The future is now :smiley:

xavierleroy commented:
I love it when an esteemed start-up feels a practical need for my crazy code :-)  Said start-up is most welcome to test this code in anger.

braibant commented:
That's nice to see. 

I am unsure about how to interpret the need for 64 Gb of RAM. I am quite sure that I had code that hit the current limit without such a beast. I will try to play a bit with this branch to understand what was going on.

xavierleroy commented:
Marshaling "compresses" certain kinds of data.  Consider a list of booleans (or of very small integers).  The marshaled representation is 2 bytes per list cell.  The in-memory representation is 3 words (24 bytes) per list cell.  Sharing detection during marshaling adds 2 words per cell (16 bytes).  So, in the end, you need 42 bytes of in-memory data to produce 2 bytes of output.  Hence, to produce 4 Gb of output, you need 84 Gb of RAM.  In contrast, if your marshaled data consists mainly of strings or of float arrays, no compression takes place and the marshaled data has about the same size as the in-memory data.

gasche commented:
I looked at the patch and saw nothing wrong.

I wondered a bit whether it would be possible to avoid the `memmove` call in the "optimistically guess 20, or resize" strategy; you could have a header and a footer, and only store the length in the header, but that (1) is arguably ugly and (2) doesn't work as well in situations where linear reads are faster than random access (not that I think there are many with current RAM hardware), and also we can expect the optimistic guess to be almost-always true for now so there is not much point in optimising the other case. Maybe it would be nice, though, to expose a `caml_output_block_64` function that always allocate the long format, for users that predict they are in this case and don't want to pay the copy.

damiendoligez commented on byterun/extern.c:
```diff
@@ -432,6 +439,10 @@ static void extern_rec(value v)
       } else if (d < 0x10000) {
         writecode16(CODE_SHARED16, d);
+#ifdef ARCH_SIXTYFOUR
+      } else if (d >= (uintnat)1 >> 32) {
```
Surely this should be a left shift, not a right shift?

xavierleroy replied:
Surely indeed.  Fixed this on the branch.

xavierleroy committed:
Typo: >> should have been <<.


git-svn-id: http://caml.inria.fr/svn/ocaml/branches/big-marshal@16425 f963ae5c-01c2-4b8c-9fe0-0dff7051ff02

pascal-cuoq commented on byterun/extern.c:
```diff
@@ -671,50 +703,68 @@ CAMLprim value caml_output_value_to_string(value v, value flags)
+  data_len = extern_value(v, flags, header, &header_len);
+  if (header_len == 20) {
+    /* Good guess! */
+    memcpy(buf, header, 20);
```
Ces deux lignes peuvent être factorisées avec l'autre branche du if, ce qui simplifie le code dans le sens où le if-then-else devient un if-then.

xavierleroy replied:
Well spotted.  Refactored as suggested.

pascal-cuoq commented on byterun/intern.c:
```diff
@@ -73,40 +73,66 @@ CAMLnoreturn_end;
+static inline uint32_t read32u(void)
+{
+  uint32_t res =
+    (intern_src[0] << 24) + (intern_src[1] << 16)
```
Si `int` fait 32 bits, ce décalage de 24 est undefined behavior quand la valeur a son bit de poids fort positionné

xavierleroy replied:
Fixed.

pascal-cuoq commented on byterun/intern.c:
```diff
@@ -73,40 +73,66 @@ CAMLnoreturn_end;
+static inline int32_t read32s(void)
+{
+  int32_t res =
+    (intern_src[0] << 24) + (intern_src[1] << 16)
```
Ici aussi

xavierleroy replied:
Fixed as well.

pascal-cuoq commented on byterun/intern.c:
```diff
@@ -686,39 +763,54 @@ CAMLexport value caml_input_value_from_malloc(char * data, intnat ofs)
-    caml_failwith("input_value_from_block: bad block length");
-  obj = input_val_from_block();
+  caml_parse_header("input_value_from_block", &h);
+  if (h.header_len + h.data_len > len)
```
I might be wrong here, but it seems to me that a maliciously crafted header might cause the path at https://github.com/ocaml/ocaml/blob/e2d0a13165b31cb06d4a7bdba3847198b32e1bd4/byterun/intern.c#L629 to be followed and cause the values for `h.header_len` and `h.data_len` to overflow when added. This would in turn lead this condition to evaluate to false when the programmer intended true, with possibly bad consequences when `input_val_from_block` is entered.

Thinking more about it, there may already be enough ways things can go wrong when inputing a malicious value that this overflow wouldn't make a difference even if real.

xavierleroy replied:
Right, input_value and co are not protected against ill-formed inputs.  This is one example that could possibly go wrong, but I'm sure there are others.

pascal-cuoq commented:
For comments, here is a tentative patch to unmarshal.ml:

http://pastebin.com/HMJRsNX2

Does anyone have ideas about testing the pair marshal-unmarshal for the newly introduced behaviors?

pascal-cuoq commented:
For comments, here is a tentative patch to unmarshal.ml. Does anyone have ideas about testing the pair marshal-unmarshal for the newly introduced behaviors?

```
commit f7ba6926a6fb8f6299dadc27afa05f1df10d5d64
Author: Pascal Cuoq <c…@trust-in-soft.com>
Date:   Tue Sep 29 18:39:42 2015 +0200

    parse new header format when new magic code, accept new 64-bit internal values. Beware: on 32-bit the correct behavior would be to reject the latter instead (not done)

diff --git a/src/libraries/datatype/unmarshal.ml b/src/libraries/datatype/unmarshal.ml
index 8a5c192..97303d5 100644
--- a/src/libraries/datatype/unmarshal.ml
+++ b/src/libraries/datatype/unmarshal.ml
@@ -36,7 +36,7 @@

 (* caml_unmarshal by Ineffable Casters *)

-(* Version 3.11.2.0 *)
+(* Version big-marshal *)


 (* Warning:
@@ -62,8 +62,6 @@ let arch_sixtyfour = Sys.word_size = 64;;
 let arch_bigendian = (Obj.magic [| 0x00002600 |] : string).[1] <> 'L';;
 let arch_float_endianness = (Obj.magic 1.23530711838574823e-307 : string).[1];;

-let intext_magic_number = "\x84\x95\xA6\xBE";;
-
 let ill_formed reason =
   let msg = "input_value: ill-formed message" in
   failwith (if false(*debug*) then Printf.sprintf "%s (%s)" msg reason else msg)
@@ -228,10 +226,14 @@ let (code_area_start, cksum) =
 ;;
 *)

-let check_const ch s msg =
-  for i = 0 to String.length s - 1 do
-    if input_char ch <> s.[i] then failwith msg;
-  done
+let check_const ch s =
+  try
+    for i = 0 to String.length s - 1 do
+      if input_char ch <> s.[i] then raise Exit
+    done;
+    false
+  with Exit -> true
+
 ;;

 (* Auxiliary functions for handling Custom blocks. *)
@@ -341,13 +343,36 @@ let rec get_structure t context =
   | Dynamic _ -> assert false
 ;;

+(* let intext_magic_number_small = "\x84\x95\xA6\xBE";;
+let intext_magic_number_big = "\x84\x95\xA6\xBF";;
+*)
+
 let input_val ch t =
   set_binary_mode_in ch true;
-  check_const ch intext_magic_number "input_value: bad object";
-  let _block_len = getword ch in
-  let num_objects = read32u ch in
-  let _size_32 = getword ch in
-  let _size_64 = getword ch in
+  let num_objects =
+    if check_const ch "\x84\x95\xA6"
+    then failwith "input_value: bad object";
+    let last_char = input_char ch in
+    match last_char with
+    | '\xBF' ->
+      (* See
+    https://github.com/ocaml/ocaml/blob/c065a0995c0a18fcc2430977dbc887766b6ffe45/byterun/extern.c#L626
+    https://github.com/ocaml/ocaml/blob/e2d0a13165b31cb06d4a7bdba3847198b32e1bd4/byterun/caml/intext.h#L43 *)
+   if check_const ch "\x00\x00\x00\x00"
+   then failwith "input_value: bad object (reserved bytes should be 0)";
+   ignore (read64s ch); (* block_len *)
+   let n = read64u ch in
+   ignore (read64s ch); (* size_64 *)
+   n
+    | '\xBE' ->
+      ignore (getword ch); (* block_len *)
+      let n = read32u ch in
+      ignore (getword ch); (* size_32 *)
+      ignore (getword ch); (* size_64 *)
+      n
+    | _ ->
+      failwith "input_value: bad object"
+  in
   let tbl = LA.make num_objects null in
   let patch = LA.make num_objects [] in

@@ -384,6 +409,9 @@ let input_val ch t =
       | 0x06 (* CODE_SHARED32 *) ->
           let ofs = read32u ch in
           read_shared stk ofs
+      | 0x14 (* CODE_SHARED64 *) ->
+          let ofs = read64u ch in
+          read_shared stk ofs  

       | 0x08 (* CODE_BLOCK32 *) ->
           let (tag, size) = readheader32 ch in
@@ -398,6 +426,9 @@ let input_val ch t =
       | 0x0A (* CODE_STRING32 *) ->
           let len = read32u ch in
           read_string stk t len
+      | 0x15 (* CODE_STRING64 *) ->
+          let len = read64u ch in
+          read_string stk t len

       | 0x0C (* CODE_DOUBLE_LITTLE *) ->
           read_double stk t readfloat_little
@@ -415,6 +446,12 @@ let input_val ch t =
       | 0x0F (* CODE_DOUBLE_ARRAY32_BIG *) ->
           let len = read32u ch in
           read_double_array stk t len readfloat_big
+      | 0x17 (* CODE_DOUBLE_ARRAY64_LITTLE *) ->
+          let len = read64u ch in
+          read_double_array stk t len readfloat_little
+      | 0x16 (* CODE_DOUBLE_ARRAY64_BIG *) ->
+          let len = read64u ch in
+          read_double_array stk t len readfloat_big

       | 0x10 (* CODE_CODEPOINTER *) ->
    assert false
```

xavierleroy committed:
Minor fixes following the discussion on pull request 224:
- intern.c: watch out for left shift overflow
- extern.c: refactor caml_output_value_to_block


git-svn-id: http://caml.inria.fr/svn/ocaml/branches/big-marshal@16522 f963ae5c-01c2-4b8c-9fe0-0dff7051ff02

gasche commented:
For the record, a change in the demarshalling behavior of empty files (induced by this PR) is discussed as [MPR#7142](http://caml.inria.fr/mantis/view.php?id=7142) -- it breaks some users' expectation and code.
